### ISSUE TITLE
Multiple source paths solution / proof of concept fixes #72, #86, #210, #270

### ISSUE THREAD

**Issue by typhonrt** [esdoc/esdoc#310]
There have been several issues opened #72, #86, #210, #270 regarding supporting multiple source (and test paths). @h13i32maru [currently in March has declined](https://github.com/esdoc/esdoc/issues/72#issuecomment-195889544) to support this feature with the claim it is too complex for ESDoc, but my proof of concept implementation hopefully proves that it works fine while maintaining backward compatibility. It took ~20 minutes to get the first implementation working then ~3 hours to make it "bullet proof". 

I have a test branch [source-globs](https://github.com/typhonjs-node-esdoc/esdoc/tree/source-globs) that seems to offer a great solution. Since I have done the proof of concept work in a branch of my fork of [ESDoc which supports Babylon](https://github.com/typhonjs-node-esdoc/esdoc) I'm posting a new issue here instead of a PR especially given that this is a potential major (but non-breaking) change. 

I'd like to open comments in support of adding this feature into ESDoc and of course get any feedback on the implementation. 

You can install it via:
`npm install "git+https://git@github.com/typhonjs-node-esdoc/esdoc.git#source-globs"`

Or including the above in `package.json`. 

@h13i32maru I now have two major enhancements for ESDoc including the switch to Babylon parser and supporting multiple source / test globs which I'd like to upstream to ESDoc. Please consider accepting them! If you can agree to quickly review and accept PRs I'll certainly take the time to submit them! In my opinion after accepting these changes a new release `0.5` of ESDoc is pertinent. 

---

The following is a description of the changes involved in the proof of concept.

I've included the `glob` npm module and treat `config.source` if it is an array then as an array of globs otherwise the string path is converted into an array with an all inclusive glob created from the string path as the single entry. Also if`config.test.source` is an array it also is treated as an array of globs, etc. 

So if you were to include three source paths the following config file applies:

```
{
   "source": ["src/**/*.js", "src2/**/*.js", "src3/subdir/*.js"],
   "destination": "docs"
}
```

It should be noted that `config.source` / `config.test.source` now holds an array of globs as a bare string path is converted into a glob and stored as the single entry in an array. There are two new config parameters which store the base working path for source and tests which are `config.sourceDirPath` and `config.test.dirPath`.  To retain backward compatibility if either `config.source` or `config.test.source` is originally a path string then those paths define `config.sourceDirPath` and `config.test.dirPath`, but the user now has an option to override and provide in the config file `config.sourceDirPath` and `config.test.dirPath` which will be used in place instead of being generated. If not provided at all then they are set to the root path `.` when a source glob array is defined.

An example of including multiple paths relative to a user defined source directory path:

```
{
   "source": ["otherpath/src/subdir1/**/*.js", "otherpath/src/subdir2/**/*.js", "otherpath/src/subdir3/**/*.js"],
   "sourceDirPath": "otherpath/src",
   "destination": "docs"
}
```

---

I added a basic sanity check in tests to just make sure things work with source glob arrays. I have tested this in a few separate repos and so far things work great for source and test loading for existing ESDoc config files and supporting the new source globs array pattern.

While two new config parameters are added (`config.sourceDirPath` and `config.test.dirPath`) they are not necessary to set in advance, so any existing ESDoc config file works as expected.  

---

**Impacts**

With any major change it's important to consider the impact it might have on ESDoc plugins. As far as I'm aware I'm the only author of an ESDoc plugin, [esdoc-plugin-jspm](https://www.npmjs.com/package/esdoc-plugin-jspm) which uses the _old hack_ of rewriting / synthesizing the `config.includes` regexp to include multiple source directories supporting JSPM managed source code. I don't believe any other existing ESDoc plugins may be impacted and supporting source globs is a much preferred direction for `esdoc-plugin-jspm` because it prevents usage of `config.includes` by overwriting it.

Given that a bare source path is rewritten to a source glob array and stored in `config.source` and `config.test.source` and the new `config.sourceDirPath` and `config.test.dirPath` entries contain the actual paths there is a potential for ESDoc plugins to be updated. This affects my plugin [esdoc-plugin-enhanced-navigation](https://www.npmjs.com/package/esdoc-plugin-enhanced-navigation), but has low impact. In that plugin I used the old `config.source` to rewrite file paths for the navigation. This can actually now be _removed_.

So, there may be a minor impact on a few ESDoc plugins, but the feature enhancement is much improved and since ESDoc is no where near `v1.0` any affected plugins (if any at all outside my efforts) can easily make the change to use `config.sourceDirPath` as necessary.

**Comment by raveclassic** [comment]
Great work!
I'll provide comments for my PR and this branch here.
> I think it would be great to include the glob handling automatically.

I'm ok with that although I don't believe it's a responsibility of this library to handle globs. But it's ok. 
The only question is about handling glob options such as `ignore`. This is very important in cases like react component demo scripts that should not be included in the doc.
Now I can handle it by filtering files firstly and then passing the list to doc generator:
```
glob.sync('./src/components/**/*.jsx', {
 ignore: ['./src/components/**/*.demo.jsx']
})
```
Of course I can still filter files firstly and then just pass the resulting list to `source`.
Maybe inclusion of some kind of `ignore` param to doc config should be considered?

**Comment by typhonrt** [comment]
>The only question is about handling glob options such as ignore.

The `config.includes` and `config.excludes` regexp is still available for use. With the _old hack_ a plugin like `esdoc-plugin-jspm` would overwrite `config.includes` to include multiple paths preventing it's usage and not fully picking up any includes / excludes usage potentially defined in any subsequently linked JSPM repos with different options in their respective ESDoc config files. Now it's possible to handle many use cases. As you mention you can also feed `config.source` with an explicit array of files.

In your example of ignoring `*.demo.jsx` you can set `config.excludes` to `\\.demo\\.jsx$`, etc

**Comment by typhonrt** [comment]
For those interested in the implementation specifics the new method, [_hydrateSourceGlob](https://github.com/typhonjs-node-esdoc/esdoc/blob/source-globs/src/ESDoc.js#L102-L150) does all the heavy lifting so to speak. The old `_walk` all inclusive directory walking has been removed as `_hydrateSourceGlob` provides a precise list of files for doc generation. 

It is invoked in `generate` [here](https://github.com/typhonjs-node-esdoc/esdoc/blob/source-globs/src/ESDoc.js#L39) and in `_generateTest` [here](https://github.com/typhonjs-node-esdoc/esdoc/blob/source-globs/src/ESDoc.js#L163).

The remainder of the changes regard patching the few locations where previously `config.source` / `config.test.source` where used as paths. Search for `config.sourceDirPath` and `config.test.dirPath` to see these points.

------

For plugin consideration. The `onHandleConfig` callback occurs before glob  / source file list creation. So plugins will have an opportunity to modify `config.source`. Be aware that in any subsequent plugin callbacks that `config.source` and `config.test.source` will be the resolved glob array, so if a bare string path is originally set just after `onHandleConfig` it will be an array with a resolved all inclusive glob string. Working paths for source paths (`config.sourceDirPath`) and test source paths (`config.test.dirPath`) are also resolved after `onHandleConfig` is invoked.

**Comment by raveclassic** [comment]
To be honest I don't see any profit in mixing both glob filtering and regexps. I believe something one should be done - either removing globs from source param in config (but still accepting arrays of paths allowing both relative and absolute) or replacing includes/excludes with ignore param accepting the same array of globs. This will make configuration clear and consistent.
Personally I don't know what would be the best decision but I do believe that smth one should be left. What do you think?

**Comment by typhonrt** [comment]
Not breaking existing functionality is most important. Any changes should be backward compatible otherwise any hope of getting @h13i32maru to accept upstream goes out the window more or less. So `includes/excludes` has to stay. 

Also consider that you are using ESDoc in a programmable context and that the CLI usage is the normal case. 

That doesn't mean you don't bring up a valid view point for the programmable case. I'm actually all for perhaps adding `config.sourceFiles` and `config.test.files` and if either of those are defined as an array of file paths then glob hydration is skipped. Also it would make sense to store the file list created from globs respectively after filtering  in `sourceFiles / test.files`. While on a roll perhaps another option `config.globOptions` could be fed into glob usage thus supporting the full range of available functionality with this path. 

All of this should be possible with full backward compatibility and greatly expand the flexibility of ESDoc. I just didn't try to "move the cheese" so to speak too far with this current effort as who knows even if it will get upstream. I still think adding all of the above and switching to Babylon for the parser makes a great 0.5 release. IE my branch is already versioned `0.5-alpha`.. :D

**Comment by raveclassic** [comment]
Yes, I agree with all you've described and also am looking forward to merging these changes to base esdoc.

All I would like to see in this great tool is a consistency of configuration. Backward compatibility for includes/excludes should be supported of course but can be easily deprecated for some range of future versions. 

I should state that my problem is solved by adding support of file path arrays. This is all related to the case when source configuration is improved with globs and I believe globs are easier to maintain and are much more common because of tools like gulp.

**Comment by typhonrt** [comment]
Just posting a success story @h13i32maru that file globs and multiple source roots works for a really large series of projects for an external user. See #316 for the full walkthrough. It really would be great to get this committed to the mainline ESDoc repo.

**Comment by raveclassic** [comment]
Guys, the solution ESDoc provides for documenting JS is really far from complete. 
We've been trying to adopt our codebase for ESDoc struggling with unsupported features and missed edge cases but at last we've ended with `typedoc` and separate `.d.ts` files. We are also really lucky that WebStorm picks up `.d.ts` to help with code inspections.

Our security policy does not allow to point to github (yep that's weird but that's it) so we need ESDoc to be published to npm. So, @typhonrt,  thank you for the effort you put to ESDoc with your fork. It's really promising and I do believe @h13i32maru will someday merge it.

Summing up for now we're with typedoc but looking forward to further development of esdoc.

**Comment by h13i32maru** [comment]
For now, I do not support multi source root. If you want to use it, please use plugin API.

rough code:
```js
exports.onHandleConfig = function(ev) {
  const sources = ev.data.config.sources
  const {source, includes, excludes} = convertSourcesToSource(sources);
  ev.data.config.source = source;
  ev.data.config.includes = includes;
  ev.data.config.excludes = excludes;
};
```

**Comment by typhonrt** [comment]
I am not sure how your example above accomplishes multi-source root usage. Do you care to enlighten? Using includes and excludes from a common root is not an adequate solution. 

Like I mentioned I have a clear proof of concept implementation of this functionality and it works fine. 

By rejecting without discussion is a main impetus for a permanent fork of ESDoc. At one point in time which is ever passing I'd have been glad to submit a PR, but decisions like this need buy in from you the single maintainer before I'd waste time submitting such a PR which would likely get tossed aside.

**Comment by raveclassic** [comment]
At least we have tried.

**Comment by raveclassic** [comment]
@typhonrt I believe that I should note that we switched to TS/typedoc completely. TS2 supports `--allowjs` flag that allows ts compiler to parse existing jsdoc in js files (yep, you don't need to mess with `.d.ts` anymore) so you don't need to rewrite your codebase. So I highly recommend you to take a look.

**Comment by IzioDev** [comment]
@typhonrt Any chance of making your multiple source work?
